packages/firestore/scripts/assertion-id-tool.ts

```diff
@@ -0,0 +1,372 @@
+              errorMessage,
+              assertionId: assertionId ?? "INVALID",
+            });
+          }
+        }
+
+        // Continue traversing down the AST
+        ts.forEachChild(node, visit);
+      };
+
+      // Start traversal from the root SourceFile node
+      visit(sourceFile);
+
+    } catch (error: any) {
+      console.error(`Error processing file ${filePath}: ${error.message}`);
+    }
+  } // End loop through filePaths
+
+  return foundCalls;
+}
+
+
+// --- Action Handlers ---
+
+function handleList(occurrences: CallSiteInfo[]): void {
+  if (occurrences.length === 0) {
+    log("No assertion ids found.");
+    return;
+  }
+
+  occurrences.sort((a, b) => a.assertionId.localeCompare(b.assertionId)).forEach((call) => {
+    console.log(
+      `ID: ${call.assertionId}; MESSAGE: ${call.errorMessage}; SOURCE: '${call.functionName}' call at ${path.relative(process.cwd(), call.fileName)}:${call.line}:${call.character}`
+    );
+  });
+}
+
+function find(occurrences: CallSiteInfo[], targetId: string | number): CallSiteInfo[] {
+  const target = typeof targetId === 'number' ? targetId.toString(16) : targetId;
+  return occurrences.filter(o => String(o.assertionId) === String(target));
+}
+
+function handleFind(occurrences: CallSiteInfo[], targetId: string | number): void {
+  const foundLocations = find(occurrences, targetId);
+
+  if (foundLocations.length === 0) {
+    log(`Assertion id "${targetId}" not found.`);
+    process.exit(1);
+  }
+
+  handleList(foundLocations);
+}
+
+function handleCheck(occurrences: CallSiteInfo[]): void {
+  if (occurrences.length === 0) {
+    log("No assertion ids found to check for duplicates.");
+    return;
+  }
+  const idCounts: { [id: string]: CallSiteInfo[] } = {};
+
+  occurrences.forEach(occ => {
+    // Count ID occurrences
+    const codeStr = String(occ.assertionId); // Use string representation as key
+    if (!idCounts[codeStr]) {
+      idCounts[codeStr] = [];
+    }
+    idCounts[codeStr].push(occ);
+
+    // validate formats
+    if (!/^0x[0-9a-f]{4}$/.test(occ.assertionId)) {
+      console.error(`Invalid assertion ID '${occ.assertionId}'. Must match /^0x[0-9a-f]{4}$/`);
+
+      const relativePath = path.relative(process.cwd(), occ.fileName);
+      console.error(`- at '${relativePath}:${occ.line}:${occ.character}`);
+    }
+  });
+
+  let duplicatesFound = false;
+  log("Checking for duplicate assertion id usage:");
+  Object.entries(idCounts).forEach(([code, locations]) => {
+    if (locations.length > 1) {
+      duplicatesFound = true;
+      console.error(`\nDuplicate assertion id "${code}" found at ${locations.length} locations:`);
+      locations.forEach(loc => {
+        const relativePath = path.relative(process.cwd(), loc.fileName);
+        console.error(`- ${relativePath}:${loc.line}:${loc.character}`);
+      });
+    }
+  });
+
+  if (!duplicatesFound) {
+    log("No duplicate assertion ids found.");
+  }
+  else {
+    process.exit(1);
```

Comment:
Consider throwing an exception rather than calling process.exit().

Reply:
On this one I prefer `process.exit(1)`. The intent is for this to fail the build with a non-zero exit code. Throwing an error could make the failure appear as if there was an issue with the script, rather than an issue with the code it is checking.